cdap-api/src/main/java/co/cask/cdap/api/messaging/MessageFetcher.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.messaging;
+
+import co.cask.cdap.api.annotation.Beta;
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+
+import java.io.IOException;
+import javax.annotation.Nullable;
+
+/**
+ * Provides access to fetch messages from the transactional messaging system.
+ */
+@Beta
+public interface MessageFetcher {
+
+  /**
+   * Fetch messages from the given topic that was published on or after the given timestamp.
+   *
+   * @param namespace namespace of the topic
+   * @param topic name of the topic
+   * @param limit maximum number of messages to fetch
+   * @param timestamp timestamp in milliseconds; the publish timestamp of the first {@link Message}
+   *                  returned by the resulting {@link CloseableIterator} will be on or after the given timestamp.
+   *                  Use {@code timestamp = 0} to poll from the first available message.
+   * @return a {@link CloseableIterator} of {@link Message}
+   * @throws IOException if failed to communicate with the messaging system.
+   * @throws IllegalArgumentException if the give topic doesn't exist or if the timestamp is negative.
```

Comment:
See my comments below in https://github.com/caskdata/cdap/pull/7250#discussion_r90916011